### PR TITLE
Add GPUShaderModuleDescriptor/sourceMap.

### PR DESCRIPTION
**Pull request by jdashg**
***
<a href="https://pr-preview.s3.amazonaws.com/jdashg/gpuweb/pull/645.html" title="Last updated on May 18, 2020, 9:10 PM UTC (cca0be7)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/gpuweb/gpuweb/645/23b35a9...jdashg:cca0be7.html" title="Last updated on May 18, 2020, 9:10 PM UTC (cca0be7)">Diff</a>

### CODE REVIEW

File: spec/index.bs [hunk 1]
```diff
@@ -2008,9 +2008,15 @@ conditions.
 <script type=idl>
 dictionary GPUShaderModuleDescriptor : GPUObjectDescriptorBase {
     required DOMString code;
+    object? sourceMap = null;
 };
 </script>
 
+{{GPUShaderModuleDescriptor/sourceMap}}, if non-null, is interpreted as a source-map-v3 format.
```
Comment (kainino0x) [inline spec/index.bs]: Dictionary members are optional by default:

```suggestion
    object sourceMap;
};
</script>

{{GPUShaderModuleDescriptor/sourceMap}}, if defined, is interpreted as a source-map-v3 format.
```

(here I'm using "defined" to mean "both set and not `undefined`". I don't know what the standard wording is.)

Comment (jdashg) [reply spec/index.bs]: "defined" sounds accurate

**Comment by Kangz** [comment]
TL;DR The combination of these two PRs #645 and #646 make source maps normative and I want to make sure we are aware of it when we make decisions on them.

While I very much agree with the sentiment in this issue that this is a good thing, I believe it will be the first time that source maps are added in an application-visible way to the Web platform. AFAIK previous usages only interact with the devtools and are not observable programmatically.

The addition of sourcemaps here and the `mapped*` members in #646 mean that the interpretation of sourcemaps by the browser is visible to the application and we now need to have a normative specification of sourcemaps and of the relationship between non-mapped and mapped source position. That's unless we say that `mapped*` are pure hints to the application and cannot be relied to be portable.

**Comment by damyanp** [comment]
If we don't make them normative then I wonder what would prevent them becoming vendor-specified opaque blobs?

**Comment by jdashg** [comment]
There's already a cross-browser spec for this, and the way I referenced it here does make it normative.
The existing spec is for ensuring that sourcemaps are cross-UA interoperable. Though there's no programatic way to validate this at the moment (I think?), there's pre-existing vested interest by UAs in ensuring cross-compatibility.
I think the main difference here is that we're making this cross-compatibility observable, but I don't think we need much more work do make this happen. I suppose we should reach out to our source-maps colleagues and check!

File: spec/index.bs [hunk 2]
```diff
@@ -2008,9 +2008,15 @@ conditions.
 <script type=idl>
 dictionary GPUShaderModuleDescriptor : GPUObjectDescriptorBase {
     required DOMString code;
+    object sourceMap;
 };
 </script>
 
+{{GPUShaderModuleDescriptor/sourceMap}}, if defined, is interpreted as a source-map-v3 format.
```
Comment (dneto0) [inline spec/index.bs:2015]: What happens if there's a sourcemap v4?  Should there be another field that specifies the version of the sourcemap?  Or would you reuquire a new GPUShaerModuleDescriptor when that time comes?

**Comment by grorg** [comment]
Discussed at the [2020-04-21 meeting](https://docs.google.com/document/d/1v6U_0yh3HQZ1rb53A4Mi_OgK9Mn6d0w9lYjDn2kkG_4/)

File: spec/index.bs (hunk 2 above)
Comment (kainino0x) [reply spec/index.bs:2015]: [from meeting] Source map spec has a version field at the top level.

**Comment by kainino0x** [comment]
As I was talking about on the call, I think the solution used by source maps for JavaScript is a good one: there's no breakable API surface that cares about what the format of the source map is. The browsers need only to support it in their dev tools. If the application wants to apply source mapping to errors, it can do so with a [userspace source map library](https://www.npmjs.com/package/source-map). They ship their source maps along with that library so there's no possibility of breakage.

Taking that approach is compatible with this PR as-is, but requires mappedLine et al to be removed from #646.

Did you happen to get feedback from your team on that approach?

**Comment by litherum** [comment]
The case we’re worried about is browser A switches to some proprietary source map format, an author develops a page in that browser using this proprietary format, someone else opens that page in Safari’s Web Inspector, sees unreadable shader contents in the relevant panel, and thinks that Safai’s dev tools are worse than Chrome’s.

**Comment by kainino0x** [comment]
But that's exactly the situation with JavaScript source maps. What makes WGSL special?

I also assume each browser would probably reuse the implementation it uses for JS for WGSL, so I really don't think WGSL source maps would diverge faster than JS source maps. Unless we are talking about some format that's somehow conceptually far better than source maps, in which case a browser's dev tools *would* actually be concretely better for supporting that new format.

**Comment by litherum** [comment]
> But that's exactly the situation with JavaScript source maps. What makes WGSL special?

Our representatives to ECMA are concerned about this too.

**Comment by grorg** [comment]
@jdashg do you think this should be discussed again at the WebGPU meeting or the WGSL meeting?

**Comment by jdashg** [comment]
For topics that straddle the API/shader boundary, I think the main group is best. In my mind the language meetings are like a sub-group that reports up to the main group, so topics that concern everyone would default to the full main CG meeting.

**Comment by kvark** [comment]
One idea that came up in a discussion with @jdashg is making this an extension. That would allow us to still test this with the CTS, and yet allow to iterate on that later. For example, if sourcemap v4 standard comes out, we can just release a new extension.

File: spec/index.bs (hunk 2 above)
Comment (jdashg) [reply spec/index.bs:2015]: I will s/is interpreted as/MAY be interpreted as/
